cpp/daal/src/algorithms/dtrees/gbt/classification/gbt_classification_predict_dense_default_batch_impl.i

```diff
@@ -178,8 +178,103 @@ public:
 protected:
     services::Status predictByAllTrees(size_t nTreesTotal, size_t nClasses, const DimType & dim);
 
-    void predictByTrees(algorithmFPType * res, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x);
-    void predictByTreesVector(algorithmFPType * val, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x);
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    using Dispetcher_t = gbt::prediction::internal::PredictDispetcher<hasUnorderedFeatures, hasAnyMissing>;
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    void predictByTrees(algorithmFPType * val, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x,
+                        const Dispetcher_t<hasUnorderedFeatures, hasAnyMissing>& dispetcher);
+    template <bool hasUnorderedFeatures, bool hasAnyMissing>
+    void predictByTreesVector(algorithmFPType * val, size_t iFirstTree, size_t nTrees, size_t nClasses, const algorithmFPType * x,
+                              const Dispetcher_t<hasUnorderedFeatures, hasAnyMissing>& dispetcher);
+
+    template <bool val>
+    struct BooleanConstant {
+        typedef BooleanConstant<val> type;
+    };
+
+    inline void updateResult(algorithmFPType * res, algorithmFPType * val, size_t iRow, size_t i, size_t nClasses, BooleanConstant<true> dispetcher) {
+        res[iRow + i] = algorithmFPType(getMaxClass(val + i * nClasses, nClasses));
+    }
+
+    inline void updateResult(algorithmFPType * res, algorithmFPType * val, size_t iRow, size_t i, size_t nClasses, BooleanConstant<false> dispetcher) {}
+
+    inline algorithmFPType* updateBuffer(algorithmFPType * val, size_t buf_shift, size_t buf_size, BooleanConstant<true> dispetcher) {
+        services::internal::service_memset_seq<algorithmFPType, cpu>(val, algorithmFPType(0), buf_size);
+        return val;
+    }
+
+    inline algorithmFPType* updateBuffer(algorithmFPType * val, size_t buf_shift, size_t buf_size, BooleanConstant<false> dispetcher) {
```

Comment:
I don't know how this is handled in the remaining part of the codebase, but seeing these two `updateBuffer` functions confuses me. We have specialization implementations for the `BooleanConstant` value and depending on that we apply the `buf_shift` or not? This is completely confusing. In the first implementation `buf_shift` isn't even used. And neither implementation uses `dispetcher` (again, typo).
Why isn't this something like

```c++
inline algorithmFPType* initAndUpdateBuffer(algorithmFPType * val, size_t buf_size);
inline algorithmFPType* updateBuffer(algorithmFPType * val, size_t buf_shift, size_t buf_size);
```

What is the use of the second function? Can't we just calculate val + buf_shift in place?